### PR TITLE
feat(orchestration-v2): import an existing Claude Code or Codex session by id, with two-way transcript sync

### PR DESCRIPTION
**Pull request by Bil0000**
Reimplements #4617 against the orchestration V2 runtime, as requested in https://github.com/pingdotgg/t3code/pull/4617#issuecomment-5132984246 — plus a safe two-way sync between the imported thread and the provider's own on-disk transcript.

## Why

T3 Code can only continue sessions it started. A Claude Code CLI session or a `codex` thread already on the machine is fully resumable, but there was no way to open it here. The V1 attempt (#4617) was closed because #2829 replaced the orchestration layer it was built on; this is the V2 reimplementation, targeting the #2829 branch.

## What

`Import session...` in the command palette takes one session id.

- **Resolve first.** `orchestration.resolveImportSession` reads the session's working directory and title before anything is created. If a project covers that directory, the thread lands there regardless of which project is open; if not, the dialog relabels to `Add project & import` and nothing is created without a second press. Unlike V1, Codex sessions resolve too (workspace + title come from the rollout file).
- **Import is all-or-nothing.** The transcript is read *before* anything is written — Claude via the agent SDK (`getSessionInfo`/`getSessionMessages`), Codex from its rollout `.jsonl` under `$CODEX_HOME/sessions` (no live app-server needed). History is appended as synthetic `message.updated`/`turn-item.updated` pairs with deterministic ids — the same mechanism as the legacy V1 thread importer — under a new `historyOrigin: "provider_import"`. There is no create-then-compensate dance: a failed import leaves nothing behind.
- **The next turn resumes the real session.** The import synthesizes a provider thread whose `nativeThreadRef` is the external session id (`strength: "strong"`; for Claude, `nativeConversationHeadRef` is set to the last transcript message so the adapter takes the `resume` path instead of colliding on `sessionId`). `ProviderTurnStartService` then resumes natively — the model keeps its full context; nothing is replayed.

## Two-way sync

- **T3 → provider comes free.** Because the thread resumes the original native session, Claude appends every T3 turn to the original `~/.claude/projects/**/<id>.jsonl`, and Codex to its own rollout.
- **Provider → T3 is a lazy sync.** On thread reads, `ensureSynced` re-reads the transcript (throttled by source mtime, serialized per thread, skipped while a run is active) and appends turns made outside T3. Idempotency and safety come from three dedupe layers: deterministic event ids, provider-native item ids already present as turn items, and exact role+text match against native run messages — so T3's own turns echoed back by the provider are never re-imported. Sync state lives in `orchestration_v2_session_imports` (migration 046).

## Verification

- 8 new server tests: pure transcript-mapper tests plus an integration test that imports a real on-disk Codex rollout into an in-memory V2 store, checks the projection (history origin, messages, provider thread binding), refuses duplicates, and round-trips the provider→T3 sync.
- 6 new `CommandPalette.logic` tests (17/17 in the file).
- `fmt`, `lint`, `typecheck` clean across contracts, server, client-runtime, web (pre-existing `@cursor/sdk`-related failures unaffected).
- **Live, against a real Claude Code session:**
  1. `claude -p "Remember this fact for later: the import canary code is FERRET-4413. Reply with exactly: OK"` in a project directory.
  2. Palette → `Import session...` → paste id → thread appears with both prior messages rendered.
  3. Ask "What is the import canary code?" — the resumed session answers **FERRET-4413**. The canary only exists in the on-disk transcript, so the real session resumed rather than the history being replayed.
  4. The question and answer from step 3 appear in the original `~/.claude/projects/**/<id>.jsonl` (T3 → provider).
  5. `claude -p --resume <id> "…second canary is BADGER-7788…"` in the terminal, reopen the thread in T3 — the CLI turn appears (provider → T3), and T3's own turns are not duplicated.

## Demo

https://drive.google.com/file/d/16rBmArbhFZyHfzLWxNY7pxJNyqt7D9hs/view?usp=sharing

## Known limits

- ~~Only plain user/assistant text is imported~~ **Solved:** tool activity now survives the import — Claude thinking blocks become reasoning items, Bash calls become command items paired with their results, Edit/Write calls become file-change items (other tools get a compact command rendering); same for Codex rollouts (reasoning summaries, function calls paired by call id). Verified live against a real session with tool use.
- ~~Turns synced from the CLI render above native T3 runs~~ **Solved:** sync-appended entries take position ordinals after the thread's current maximum, so CLI turns made after T3 turns render chronologically. Verified live (CLI turn after a native T3 turn lands at the bottom).
- ~~`$CODEX_HOME` only~~ **Solved:** rollout discovery resolves the codex instance's configured home from its resolved home layout first, then `$CODEX_HOME`, then `~/.codex`.
- Codex live resume remains unexercised end-to-end (no Codex auth on the test machine); transcript parsing, resolve, import and sync are covered by the integration test against a real rollout file, and resume goes through the existing `thread/resume` adapter path.

> [!NOTE]
> ### Add import and two-way sync of Claude Code and Codex sessions into orchestration-v2 threads
> - Adds a full import workflow (`SessionImportService`) that reads an external Claude Code or Codex session, maps its transcript to deterministic `OrchestrationV2` events, and stores it as a new thread with `historyOrigin: 'provider_import'`.
> - Adds incremental sync (`ensureSynced`) that detects provider-side conversation added after import, backfills command outputs, and is called automatically on `getThreadProjection` and `subscribeThread`.
> - Adds two new WS RPC endpoints (`resolveImportSession`, `importSession`) with auth scopes and client-runtime command functions, plus an `ImportSessionDialog` UI accessible from the command palette.
> - Extends `ProjectionStore` ordinal logic so that provider-imported runless items are only included in history up to the correct run band when forking.
> - Adds migration 046 creating the `orchestration_v2_session_imports` tracking table with a unique index on `(driver, external_id)`.
> - Risk: `getThreadProjection` and `subscribeThread` now perform a file-system or network read (`ensureSynced`) on every call for imported threads; failures are swallowed with a warning but add latency.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized d265cee.</sup>
> 
>

### CODE REVIEW

**Comment by Bil0000** [comment]
Pushed `f00d38658` addressing three of the four known limits:

- **Rich import** — tool calls, thinking/reasoning, and file edits from the original session now import as command/reasoning/file-change turn items (Claude and Codex), with tool outputs paired to their calls. Message ids are unchanged, so threads imported before this commit sync cleanly.
- **Chronological sync** — transcript entries synced after the import now take position ordinals past the thread's current maximum, so CLI turns made after T3 turns render in order instead of inside the history block.
- **Codex home** — rollout discovery uses the instance's resolved home layout (falling back to `$CODEX_HOME`, then `~/.codex`).

All verified live against real Claude Code sessions (tool-use session import renders the executed command with output; a CLI turn added after a native T3 turn appears at the bottom of the thread). 10 server tests + 6 palette tests pass; typecheck/lint/fmt clean.

The one remaining limit is Codex live resume, which needs a machine with Codex auth — covered by unit/integration tests up to the `thread/resume` adapter boundary.

File: apps/server/src/orchestration-v2/SessionImportService.ts [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+
+export class SessionImportService extends Context.Service<
+  SessionImportService,
+  SessionImportServiceShape
+>()("t3/orchestration-v2/SessionImportService") {}
+
+const importError = (message: string, cause?: unknown) =>
+  cause === undefined
+    ? new OrchestrationV2ImportSessionError({ message })
+    : new OrchestrationV2ImportSessionError({ message, cause });
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/server/src/orchestration-v2/SessionImportService.ts]: Fixed in `3eabbcfc6` — the `importError` helper is gone; `OrchestrationV2ImportSessionError` is constructed at each failure boundary with its structured fields.

File: packages/contracts/src/orchestrationV2.ts [hunk 2]
```diff
@@ -2318,6 +2324,42 @@ export const OrchestrationV2ThreadLaunchResult = Schema.Struct({
...
+});
+export type OrchestrationV2ImportSessionResult = typeof OrchestrationV2ImportSessionResult.Type;
+
+export class OrchestrationV2ImportSessionError extends Schema.TaggedErrorClass<OrchestrationV2ImportSessionError>()(
+  "OrchestrationV2ImportSessionError",
+  {
+    message: TrimmedNonEmptyString,
+    cause: Schema.optional(Schema.Defect()),
+  },
+) {}
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply packages/contracts/src/orchestrationV2.ts:2376]: Fixed in `3eabbcfc6` — the error now carries a `reason` discriminator (`session_not_found` / `already_imported` / `workspace_mismatch` / `empty_transcript` / `provider_unsupported` / `project_missing` / `read_failed` / `store_failed`) plus the `externalId`. `message` is kept as a field rather than a derived getter because several messages embed context the structure doesn't carry (workspace paths, provider names) and the web dialog surfaces it verbatim.

File: apps/server/src/orchestration-v2/SessionImportService.ts [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+   * No-op for threads that were not created by an import. Never fails the
+   * caller: sync problems are logged and surfaced via the imports table.
+   */
+  readonly ensureSynced: (threadId: ThreadId) => Effect.Effect<void>;
+}
+
+export class SessionImportService extends Context.Service<
+  SessionImportService,
+  SessionImportServiceShape
+>()("t3/orchestration-v2/SessionImportService") {}
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/server/src/orchestration-v2/SessionImportService.ts]: Fixed in `3eabbcfc6` — the interface is declared inline in the `Context.Service` call and the implementations are typed via `SessionImportService["Service"][...]`.

File: apps/server/src/orchestration-v2/SessionImportService.ts [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+ * `$CODEX_HOME/sessions/**` (`~/.codex/sessions` by default). The sessions
+ * directory is shared between the user's own Codex CLI and T3's Codex
+ * app-server homes, so a thread started in either is found here.
+ */
+// ponytail: honors CODEX_HOME only; a codex instance configured with a custom
+// homePath would need that instance's CodexHomeLayout threaded in instead.
+const makeReadCodexRollout = (deps: {
+  readonly fileSystem: FileSystem.FileSystem;
+  readonly path: Path.Path;
+}) =>
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/server/src/orchestration-v2/SessionImportService.ts]: Fixed in `3eabbcfc6` — `FileSystem`/`Path` are acquired from the environment inside `make` and `readCodexRollout` is a closure over them; the `makeReadCodexRollout` constructor-injection wrapper is gone.

File: apps/server/src/orchestration-v2/SessionImportService.ts [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+      FROM orchestration_v2_session_imports
+      WHERE thread_id = ${threadId}
+      LIMIT 1
+    `;
+    const row = rows[0];
+    if (row === undefined) {
+      nonImportedThreadIds.add(threadId);
+      return;
+    }
+    const driver = row.driver;
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/server/src/orchestration-v2/SessionImportService.ts:634]: Fixed in `3eabbcfc6` — the import's write phase now runs under the same per-thread lock `ensureSynced` uses and clears the negative-cache entry after inserting the imports row, so a concurrent sync can no longer permanently cache the thread as non-imported. The negative cache is kept for the true-negative case (it guards most thread reads).

File: apps/web/src/components/ImportSessionDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        return stopImporting(
+          failureMessage(createResult, `Could not add ${session.workspaceRoot} as a project.`),
+        );
+      }
+      projectId = createdProjectId;
+      session = { ...session, projectId: createdProjectId };
+      setResolved(session);
+    }
+
+    const importResult = await importSession({
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/web/src/components/ImportSessionDialog.tsx:151]: Fixed in `3eabbcfc6` — when the import half of "Add project & import" fails, the dialog deletes the project it just created (best effort) and re-arms the two-phase flow, so a retry doesn't leave orphan projects behind.

File: apps/server/src/orchestration-v2/ProjectionStore.ts [hunk 7]
```diff
@@ -725,7 +725,7 @@ export function isTurnItemAtOrBeforeRun(input: {
   readonly sourceRunOrdinal: number;
 }): boolean {
   if (input.itemRunId === null) {
-    return input.historyOrigin === "v1_import";
+    return input.historyOrigin === "v1_import" || input.historyOrigin === "provider_import";
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/server/src/orchestration-v2/ProjectionStore.ts]: Fixed in `3eabbcfc6` — `isTurnItemAtOrBeforeRun` now takes the item's position ordinal and gates `provider_import` runless items with `ordinal < (sourceRunOrdinal + 1) * 1_000_000`. The initial import lives in the pre-run 0-band and synced entries take ordinals inside the run band they chronologically followed, so a fork from an earlier run no longer inherits later CLI turns. Covered by new unit cases in `ProjectionStore.test.ts`.

File: apps/server/src/orchestration-v2/SessionImportService.ts [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+    const { fileSystem, path } = deps;
+    const home = process.env["CODEX_HOME"]?.trim();
+    const root = path.join(
+      home !== undefined && home.length > 0 ? home : path.join(NodeOS.homedir(), ".codex"),
+      "sessions",
+    );
+    const names = yield* fileSystem
+      .readDirectory(root, { recursive: true })
+      .pipe(Effect.orElseSucceed(() => [] as Array<string>));
+    const relative = names.find((name) => name.endsWith(`${externalId}.jsonl`));
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/server/src/orchestration-v2/SessionImportService.ts]: Fixed in `3eabbcfc6` — the lookup matches the full `-${externalId}.jsonl` segment.

File: apps/server/src/orchestration-v2/SessionImportService.ts [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+          index,
+          sourceId: entry.sourceId,
+        })}`;
+        if (existingEventIds.has(eventId)) return false;
+        // Turns T3 itself drove are already in the thread as native run
+        // items; the provider transcript echoes them back. Match by native
+        // id first, then by exact role+text as a backstop for entries whose
+        // native ids the adapter did not record.
+        if (knownNativeIds.has(entry.sourceId)) return false;
+        if (knownRunMessages.has(`${entry.role}\n${entry.text.trim()}`)) return false;
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/server/src/orchestration-v2/SessionImportService.ts]: Fixed in `3eabbcfc6` — the role+text backstop is now a multiset keyed on run messages: each T3 run message absorbs exactly one transcript echo, so a genuinely repeated prompt/answer with an unrecorded native id still imports.

File: apps/server/src/orchestration-v2/SessionImportService.ts [hunk 10]
```diff
@@ -0,0 +1,631 @@
...
+          type: "provider-thread.updated",
+          threadId,
+          driver: providerDriver,
+          providerInstanceId: input.modelSelection.instanceId,
+          occurredAt: now,
+          payload: providerThread,
+        },
+      ];
+
+      yield* insertPositions(threadId, batch.positions).pipe(
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/server/src/orchestration-v2/SessionImportService.ts]: Fixed in `3eabbcfc6` — `insertPositions`, the event batch write and the imports-table insert now run in a single `sql.withTransaction` (under the per-thread lock), so the import commits or rolls back atomically and a retry after failure is possible.

File: apps/server/src/orchestration-v2/SessionImportService.ts [hunk 11]
```diff
@@ -0,0 +1,677 @@
...
+      return undefined;
+    }
+    const rolloutPath = path.isAbsolute(relative) ? relative : path.join(root, relative);
+    const readFailure = (cause: unknown) =>
+      importError(`Could not read the transcript for Codex thread '${externalId}'.`, cause);
+    const content = yield* fileSystem
+      .readFileString(rolloutPath)
+      .pipe(Effect.mapError(readFailure));
+    const stat = yield* fileSystem.stat(rolloutPath).pipe(Effect.mapError(readFailure));
+    const modifiedAtMs = Option.getOrUndefined(Option.map(stat.mtime, (mtime) => mtime.getTime()));
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/server/src/orchestration-v2/SessionImportService.ts]: Fixed in `3eabbcfc6` — `stat` runs before `readFileString`, so an append between the two operations leaves the recorded mtime older than the content and the next sync re-reads.

File: apps/server/src/orchestration-v2/SessionImportService.ts [hunk 12]
```diff
@@ -0,0 +1,677 @@
...
+  Effect.fnUntraced(function* (externalId: string, instanceHome: string | undefined) {
+    const { fileSystem, path } = deps;
+    const envHome = process.env["CODEX_HOME"]?.trim();
+    const home =
+      instanceHome ??
+      (envHome !== undefined && envHome.length > 0
+        ? envHome
+        : path.join(NodeOS.homedir(), ".codex"));
+    const root = path.join(home, "sessions");
+    const names = yield* fileSystem
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/server/src/orchestration-v2/SessionImportService.ts]: Fixed in `3eabbcfc6` — only a `NotFound` from `readDirectory` falls back to an empty listing; any other filesystem failure maps to an `OrchestrationV2ImportSessionError` with `reason: "read_failed"` instead of reporting the session as missing.

File: apps/server/src/orchestration-v2/SessionImportService.ts [hunk 13]
```diff
@@ -0,0 +1,778 @@
...
+    }
+    const driver = row.driver;
+    if (!isImportableSessionDriver(driver)) {
+      return;
+    }
+    const projection = yield* projectionStore.getThreadProjection(threadId);
+    // Never interleave a sync with a live provider turn: the transcript is
+    // mid-write and the projection is about to receive the same content as
+    // native run items.
+    const hasActiveRun = projection.runs.some(
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/server/src/orchestration-v2/SessionImportService.ts:642]: Fixed in `50525692a` — the run-state check is repeated inside the write transaction immediately before the event write. If a run went active in the window, the sync aborts without writing and without updating the imports row, so the next thread read retries once the run settles. (Sharing the lock with run creation would couple the import service into the turn-start path; the transactional re-check keeps the mutual exclusion local.)

File: apps/server/src/orchestration-v2/SessionImportService.ts [hunk 14]
```diff
@@ -0,0 +1,797 @@
...
+        AND aggregate_kind = 'thread'
+        AND stream_id = ${threadId}
+        AND event_id LIKE ${`${SESSION_IMPORT_EVENT_PREFIX}:%`}
+    `;
+    const existingEventIds = new Set(existingEventRows.map((existing) => existing.event_id));
+
+    const missing = source.entries
+      .map((entry, index) => ({ entry, index }))
+      .filter(({ entry, index }) => {
+        if (existingEventIds.has(importedEntryEventId({ driver, threadId, index, entry }))) {
```
[the remark this answers is not in the document]
Comment (Bil0000) [reply apps/server/src/orchestration-v2/SessionImportService.ts:693]: Fixed in `d265cee48` — sync now detects imported command items whose output has appeared in the transcript since they were written and re-emits them with the same item id and ordinal under a deterministic revision event id, so the projector upserts the completed output in place. Covered by a new integration-test scenario (function_call synced without its output, output appended later, next sync backfills it at the same ordinal).

**Comment by YoungPhlo** [comment]
Not exactly a helpful comment but I am so excited to see this land and can test on Win11/Linux/macOS for Codex and Claude Code if at all helpful, will otherwise stay out of the way - just really excited.

**Comment by juliusmarminge** [comment]
Closing as part of the open-PR backlog sweep (wave 1).

**Reason:** Abandoned draft XXL rebuild of closed #4617; import path moved on
**Related:** #8066 / #10386

Reopen if this is still wanted and you’re willing to rebase onto current `main`.
